File: frontend/lib/types/homework.ts
export type HomeworkStatus = 'ASSIGNED' | 'IN_PROGRESS' | 'SUBMITTED' | 'GRADED' | 'OVERDUE'; // CÁC TRẠNG THÁI CỦA BÀI TẬP
export type HomeworkPriority = 'LOW' | 'MEDIUM' | 'HIGH'; // MỨC ĐỘ ƯU TIÊN CỦA BÀI TẬP

export interface Homework {
  id: number; // ID BÀI TẬP
  studentId: number; // ID HỌC SINH ĐƯỢC GIAO
  studentName: string; // TÊN HỌC SINH
  sessionRecordId?: number; // LIÊN KẾT VỚI BUỔI HỌC CỤ THỂ (NẾU CÓ)
  title: string; // TIÊU ĐỀ BÀI TẬP
  description?: string; // MÔ TẢ CHI TIẾT YÊU CẦU
  dueDate: string; // HẠN CHÓT NỘP BÀI (DEADLINE)
  status: HomeworkStatus; // TRẠNG THÁI HIỆN TẠI
  priority: HomeworkPriority; // ĐỘ ƯU TIÊN
  attachmentUrls: string[]; // DANH SÁCH FILE ĐỀ BÀI DO GIA SƯ ĐÍNH KÈM
  tutorNotes?: string; // GHI CHÚ CỦA GIA SƯ KHI GIAO BÀI
  submittedAt?: string; // THỜI ĐIỂM HỌC SINH NỘP BÀI
  submissionUrls: string[]; // DANH SÁCH FILE BÀI LÀM CỦA HỌC SINH
  submissionNotes?: string; // GHI CHÚ CỦA HỌC SINH KHI NỘP BÀI
  score?: number; // ĐIỂM SỐ (SAU KHI CHẤM)
  feedback?: string; // NHẬN XÉT CỦA GIA SƯ (SAU KHI CHẤM)
  gradedAt?: string; // THỜI ĐIỂM HOÀN THÀNH CHẤM ĐIỂM
  createdAt: string; // NGÀY TẠO BÀI TẬP
  updatedAt: string; // NGÀY CẬP NHẬT GẦN NHẤT
  isOverdue: boolean; // KIỂM TRA ĐÃ QUÁ HẠN HAY CHƯA
  daysUntilDue?: number; // SỐ NGÀY CÒN LẠI ĐẾN HẠN
}

export interface HomeworkRequest {
  studentId: number; // ID HỌC SINH NHẬN BÀI
  sessionRecordId?: number; // ID BUỔI HỌC LIÊN QUAN
  title: string; // TIÊU ĐỀ BÀI TẬP MỚI
  description?: string; // NỘI DUNG GIAO BÀI
  dueDate: string; // THỜI HẠN PHẢI NỘP
  priority: HomeworkPriority; // THIẾT LẬP ĐỘ ƯU TIÊN
  tutorNotes?: string; // GHI CHÚ CỦA GIA SƯ
  attachmentUrls?: string[]; // DANH SÁCH FILE TÀI LIỆU
}

export interface HomeworkStats {
  totalHomeworks: number; // TỔNG SỐ BÀI TẬP ĐÃ GIAO
  assignedCount: number; // SỐ BÀI CHƯA LÀM
  inProgressCount: number; // SỐ BÀI ĐANG THỰC HIỆN
  submittedCount: number; // SỐ BÀI CHỜ CHẤM ĐIỂM
  gradedCount: number; // SỐ BÀI ĐÃ CHẤM XONG
  overdueCount: number; // SỐ BÀI ĐÃ QUÁ HẠN
  upcomingCount: number; // SỐ BÀI SẮP ĐẾN HẠN
  averageScore: number; // ĐIỂM SỐ TRUNG BÌNH CỦA HỌC SINH
}

export interface HomeworkSubmissionRequest {
  submissionNotes?: string; // GHI CHÚ CỦA HỌC SINH KHI NỘP
  submissionUrls?: string[]; // CÁC ĐƯỜNG DẪN FILE BÀI LÀM
}

export interface HomeworkGradingRequest {
  score: number; // ĐIỂM SỐ CHO BÀI TẬP (0-10)
  feedback?: string; // NHẬN XÉT CHI TIẾT CỦA GIA SƯ
}

export interface HomeworkStatusUpdateRequest {
  status: HomeworkStatus; // CẬP NHẬT TRẠNG THÁI MỚI CHO BÀI TẬP
}

File: backend/backend/src/main/java/com/tutor_management/backend/modules/homework/Homework.java
package com.tutor_management.backend.modules.homework;

import com.tutor_management.backend.modules.finance.SessionRecord;
import com.tutor_management.backend.modules.student.Student;
import jakarta.persistence.*;
import lombok.*;
import java.time.LocalDateTime;

@Entity
@Table(name = "homeworks")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Homework {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "student_id", nullable = false)
    private Student student;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "session_record_id")
    private SessionRecord sessionRecord; // Bài tập thuộc buổi học nào

    @Column(nullable = false)
    private String title;

    @Column(length = 2000)
    private String description;

    @Column(nullable = false)
    private LocalDateTime dueDate; // Hạn nộp

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private HomeworkStatus status; // ASSIGNED, IN_PROGRESS, SUBMITTED, GRADED, OVERDUE

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private HomeworkPriority priority; // LOW, MEDIUM, HIGH

    @Column(length = 1000)
    private String attachmentUrls; // URLs từ Cloudinary, phân cách bởi dấu phẩy

    @Column(length = 2000)
    private String tutorNotes; // Ghi chú của gia sư khi giao bài

    // Submission info
    private LocalDateTime submittedAt;

    @Column(length = 1000)
    private String submissionUrls; // File học sinh nộp

    @Column(length = 2000)
    private String submissionNotes; // Ghi chú của học sinh khi nộp

    // Grading info
    private Integer score; // Điểm số (0-100)

    @Column(length = 2000)
    private String feedback; // Nhận xét của gia sư

    private LocalDateTime gradedAt;

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        if (status == null) {
            status = HomeworkStatus.ASSIGNED;
        }
        if (priority == null) {
            priority = HomeworkPriority.MEDIUM;
        }
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();

        // Auto update status to OVERDUE if past due date and not submitted
        if (dueDate != null && LocalDateTime.now().isAfter(dueDate)
                && status == HomeworkStatus.ASSIGNED) {
            status = HomeworkStatus.OVERDUE;
        }
    }

    // Enums
    public enum HomeworkStatus {
        ASSIGNED,      // Đã giao
        IN_PROGRESS,   // Đang làm
        SUBMITTED,     // Đã nộp
        GRADED,        // Đã chấm
        OVERDUE        // Quá hạn
    }

    public enum HomeworkPriority {
        LOW,
        MEDIUM,
        HIGH
    }
}
